psydac/api/grid.py

```diff
@@ -2,25 +2,71 @@
 from psydac.core.bsplines          import basis_ders_on_quad_grid
 from psydac.fem.splines            import SplineSpace
 from psydac.fem.tensor             import TensorFemSpace
 from psydac.fem.vector             import ProductFemSpace
+
+def get_points_weights(spaces, axis, e):
+    for s in spaces:
+        if e in s.quad_grids[axis].indices:
+            i = np.where(s.quad_grids[axis].indices==e)[0][0]
+            return s.quad_grids[axis].points[i:i+1], s.quad_grids[axis].weights[i:i+1]
 #==============================================================================
 class QuadratureGrid():
-    def __init__( self, V , axis=None, ext=None):
+    def __init__( self, V , axis=None, ext=None, trial_space=None):
+
+        n_elements          = []
+        indices             = []
+        local_element_start = []
+        local_element_end   = []
+        points              = []
+        weights             = []
 
         if isinstance(V, ProductFemSpace):
-            V = V.spaces[0]
-
-        self._fem_grid            = V.quad_grids
-        self._n_elements          = [g.num_elements        for g in V.quad_grids]
-        self._local_element_start = [g.local_element_start for g in V.quad_grids]
-        self._local_element_end   = [g.local_element_end   for g in V.quad_grids]
-        self._points              = [g.points              for g in V.quad_grids]
-        self._weights             = [g.weights             for g in V.quad_grids]
+            V1 = V.spaces[0]
+            spaces = list(V.spaces)
+        else:
+            V1 = V
+            spaces = [V]
+
+        if trial_space  and not isinstance(trial_space, ProductFemSpace):
+            spaces.append(trial_space)
+        elif isinstance(trial_space, ProductFemSpace):
+            spaces = spaces + list(trial_space.spaces)
+
+        for i in range(len(V1.spaces)):
+
+            indices.append(reduce(np.union1d,[s.quad_grids[i].indices for s in spaces]))
+            local_element_start.append(V1.quad_grids[i].local_element_start)
+            local_element_end.append  (V1.quad_grids[i].local_element_end)
+            points.append(V1.quad_grids[i].points)
+            weights.append(V1.quad_grids[i].weights)
+
+            for e in np.setdiff1d(indices[-1], V1.quad_grids[i].indices):
+                if e<V1.quad_grids[i].indices[0]:
+                    local_element_start[-1] +=1
+                    local_element_end  [-1] +=1
+                    p,w = get_points_weights(spaces, i, e)
+                    points[-1]  = np.concatenate((p, points[-1]))
+                    weights[-1] = np.concatenate((w, weights[-1]))
+                elif e>V1.quad_grids[i].indices[-1]:
+                    p,w = get_points_weights(spaces, i, e)
+                    points[-1]  = np.concatenate((points[-1],p))
+                    weights[-1] = np.concatenate((weights[-1],w))
+                else:
+                    raise ValueError("Could not contsruct indices")
+
```

Comment:
Are you constructing a quadrature grid that works for the product of any two spaces? If so, is this a bug-fix?

Can this construction handle our quadrature rule for histopolation? I think we had a problem as an histopolation interval could contain some knots.

Reply:
yes it's used for any two spaces and yes it's a bug fix.
I'm not familiar with the problem in the histopolation, don't we use the same spaces always in the histopolation case?

Reply:
OK then, if this is a bug fix it is important to add a point about it to the PR description!

Indeed there is only one space involved in the histopolation, so the issue there was a bit different: we were performing a simple Gauss quadrature within each interval, which was defined by the Greville points of the parent space.

Since the histopolation was used as part of a projection operator, it should be exact when applied to a spline of the space you are projecting to. But unfortunately this was not the case, because quadrature intervals could contain knots and Gauss quadrature does not work for piecewise polynomials!

I hope I could explain myself...

Reply:
I see, but this class is only used for the API, so we don't use this part of the code to calculate the histopolation matrix